File: mobile/lib/utils/space_album_folders.dart
import 'package:immich_mobile/domain/models/space_album.model.dart';
import 'package:immich_mobile/domain/models/space_album_folder.model.dart';

class FolderNode {
  FolderNode(this.folder) : children = [];
  final SpaceAlbumFolder folder;
  final List<FolderNode> children;
}

class FolderSearchHit {
  const FolderSearchHit(this.album, this.path);
  final SpaceAlbum album;
  final List<String> path;
}

const _previewLimit = 4;

/// Every function here is TOTAL: it must never throw and never loop, on a dangling parentId, a
/// self-reference, or a cycle. Sync makes no ordering guarantee, so those are normal mid-sync
/// states, not corrupt input.
Map<String, SpaceAlbumFolder> _byId(List<SpaceAlbumFolder> folders) => {for (final f in folders) f.id: f};

/// Walks parent links upward with a visited guard, so any cycle terminates.
List<SpaceAlbumFolder> _ancestors(Map<String, SpaceAlbumFolder> index, String folderId) {
  final chain = <SpaceAlbumFolder>[];
  final seen = <String>{};
  var current = index[folderId];
  while (current != null && seen.add(current.id)) {
    chain.add(current);
    final parentId = current.parentId;
    current = parentId == null ? null : index[parentId];
  }
  return chain;
}

/// A folder is a root when it has no parent, its parent is absent (not yet synced), or it is a
/// member of a cycle — including the length-1 case, a self-reference. Cycle membership must be
/// decided HERE, not patched after the fact: a folder hanging below a cycle member (not itself in
/// the cycle) still needs to nest normally under it, and only classifying at construction time
/// keeps that distinction intact.
bool _isRoot(Map<String, SpaceAlbumFolder> index, SpaceAlbumFolder f) {
  final parentId = f.parentId;
  if (parentId == null || !index.containsKey(parentId)) return true;
  return _isCycleMember(index, f);
}

/// True when walking the parent chain up from [f] eventually arrives back at [f] itself. Guarded
/// with a `seen` set keyed by id, so any cycle length terminates the walk in at most
/// `index.length` steps instead of looping.
bool _isCycleMember(Map<String, SpaceAlbumFolder> index, SpaceAlbumFolder f) {
  final seen = <String>{};
  var current = index[f.parentId];
  while (current != null && seen.add(current.id)) {
    if (current.id == f.id) return true;
    final parentId = current.parentId;
    current = parentId == null ? null : index[parentId];
  }
  return false;
}

List<FolderNode> buildFolderTree(List<SpaceAlbumFolder> folders) {
  final index = _byId(folders);
  final nodes = {for (final f in folders) f.id: FolderNode(f)};
  final roots = <FolderNode>[];

  for (final f in folders) {
    final node = nodes[f.id]!;
    if (_isRoot(index, f)) {
      roots.add(node);
    } else {
      nodes[f.parentId]!.children.add(node);
    }
  }

  // No unreachable-node promotion pass is needed here (there used to be one; it was removed —
  // see the mobile Task 6 review). `_isRoot` now classifies every cycle member, including
  // self-references, as a root at construction time. Every other folder's parent chain is
  // therefore guaranteed to terminate at a root (null parent, a dangling parent, or a cycle
  // member) within `folders.length` steps, so it gets attached under a real node instead of being
  // orphaned. A regression that reintroduced an unreachable node would show up as a missing id in
  // `buildFolderTree`'s output or a cycle in `.children` — both are asserted by the T-04 group in
  // `space_album_folders_test.dart`.
  return roots;
}

List<SpaceAlbumFolder> folderPath(List<SpaceAlbumFolder> folders, String? folderId) {
  if (folderId == null) return const [];
  return _ancestors(_byId(folders), folderId).reversed.toList();
}

({List<SpaceAlbumFolder> folders, List<SpaceAlbum> albums}) folderContents(
  List<SpaceAlbumFolder> folders,
  List<SpaceAlbum> albums,
  String? folderId,
) {
  final index = _byId(folders);
  if (folderId != null && !index.containsKey(folderId)) {
    return (folders: const [], albums: const []);
  }

  // T-08: an album whose folderId names a folder we have not synced yet falls back to the ROOT,
  // rather than being hidden at every level. On mobile this is routine, not exotic.
  String? effectiveFolder(SpaceAlbum a) => a.folderId != null && index.containsKey(a.folderId) ? a.folderId : null;

  return (
    folders: folders
        .where((f) => folderId == null ? _isRoot(index, f) : f.parentId == folderId && f.id != folderId)
        .toList(),
    albums: albums.where((a) => effectiveFolder(a) == folderId).toList(),
  );
}

Set<String> _subtreeIds(List<SpaceAlbumFolder> folders, String folderId) {
  final childrenByParent = <String, List<String>>{};
  for (final f in folders) {
    if (f.parentId != null && f.parentId != f.id) {
      childrenByParent.putIfAbsent(f.parentId!, () => []).add(f.id);
    }
  }
  final ids = <String>{};
  final stack = [folderId];
  while (stack.isNotEmpty) {
    final current = stack.removeLast();
    if (!ids.add(current)) continue;
    stack.addAll(childrenByParent[current] ?? const []);
  }
  return ids;
}

List<SpaceAlbum> _albumsInSubtree(List<SpaceAlbumFolder> folders, List<SpaceAlbum> albums, String folderId) {
  final ids = _subtreeIds(folders, folderId);
  return albums.where((a) => a.folderId != null && ids.contains(a.folderId)).toList();
}

int recursiveAlbumCount(List<SpaceAlbumFolder> folders, List<SpaceAlbum> albums, String folderId) =>
    _albumsInSubtree(folders, albums, folderId).length;

List<SpaceAlbum> folderPreviewAlbums(List<SpaceAlbumFolder> folders, List<SpaceAlbum> albums, String folderId) {
  // Filter, then sort, then take — in that exact order.
  //
  // Filter: a null thumbnailAssetId means the album has no space-visible cover (it's empty, or
  // its only asset isn't visible in this space). Emitting it renders a broken tile — the exact
  // bug the server-side COALESCE prevents. This has to happen before the take: the take can only
  // ever shrink what a caller later filters, never recover a good cover it already discarded, so
  // filtering after the take can silently return an all-blank collage while good covers sit
  // further down the list.
  //
  // Sort (after filtering) then take, never take then sort: take-then-sort returns an arbitrary
  // subset — the web implementation shipped that bug once and it is invisible unless the newest
  // album sits late in the list.
  //
  // The comparator also breaks ties on `id`: List.sort() is not guaranteed stable above ~32
  // elements, so relying on input order to break `updatedAt` ties (e.g. a bulk import landing many
  // albums in the same tick) could reshuffle the 4-item preview between rebuilds. Ordering fully
  // by (updatedAt desc, id asc) makes the result independent of the sort algorithm's stability.
  final withCovers = _albumsInSubtree(folders, albums, folderId).where((a) => a.thumbnailAssetId != null).toList()
    ..sort(_byRecencyThenId);
  return withCovers.take(_previewLimit).toList();
}

/// An album's newest photo, falling back to its own `updatedAt` when it has none.
///
/// Deliberately the SAME key web sorts previews by (`endDate ?? updatedAt`, see
/// `space-album-folders.ts`). This used to be `updatedAt` alone, which meant the two clients
/// picked different covers for the same folder — a folder full of old holiday photos that was
/// merely re-synced recently would jump to the front of the collage on mobile but not on web.
DateTime _recencyOf(SpaceAlbum album) => album.endDate ?? album.updatedAt;

int _byRecencyThenId(SpaceAlbum a, SpaceAlbum b) {
  final byDate = _recencyOf(b).compareTo(_recencyOf(a));
  return byDate != 0 ? byDate : a.id.compareTo(b.id);
}

/// A folder's recursive album count and its preview albums, computed together.
class FolderSummary {
  const FolderSummary({required this.albumCount, required this.previewAlbums});

  final int albumCount;
  final List<SpaceAlbum> previewAlbums;

  static const empty = FolderSummary(albumCount: 0, previewAlbums: []);
}

/// Every folder's summary, in one pass over the space.
///
/// [recursiveAlbumCount] and [folderPreviewAlbums] are each O(folders + albums) on their own:
/// both rebuild the parent index and re-scan every album. Calling them from a grid's
/// `itemBuilder` — once per folder tile, and again for every tile whenever the parent rebuilds —
/// makes scrolling a folder list cost O(tiles x (folders + albums)) per frame. Building the two
/// indexes once and walking only each folder's own subtree makes it O(folders x depth + albums)
/// for the whole level, with depth capped at 10 by the server.
///
/// Cycle and dangling-parent behaviour is identical to the single-folder functions: a
/// self-reference is not a child of itself, and the `seen` set terminates any longer cycle.
/// `space_album_folders_test.dart` pins that equivalence folder-by-folder so the two cannot
/// drift apart.
Map<String, FolderSummary> buildFolderSummaries(List<SpaceAlbumFolder> folders, List<SpaceAlbum> albums) {
  final childrenByParent = <String, List<String>>{};
  for (final f in folders) {
    final parentId = f.parentId;
    if (parentId != null && parentId != f.id) {
      childrenByParent.putIfAbsent(parentId, () => []).add(f.id);
    }
  }

  final albumsByFolder = <String, List<SpaceAlbum>>{};
  for (final a in albums) {
    final folderId = a.folderId;
    if (folderId == null) continue;
    albumsByFolder.putIfAbsent(folderId, () => []).add(a);
  }

  final summaries = <String, FolderSummary>{};
  final seen = <String>{};
  final stack = <String>[];

  for (final folder in folders) {
    seen.clear();
    stack
      ..clear()
      ..add(folder.id);

    var albumCount = 0;
    final withCovers = <SpaceAlbum>[];

    while (stack.isNotEmpty) {
      final current = stack.removeLast();
      if (!seen.add(current)) continue;

      for (final a in albumsByFolder[current] ?? const <SpaceAlbum>[]) {
        albumCount++;
        // Filter before the take, exactly as folderPreviewAlbums does: a null cover renders a
        // broken tile, and taking first would discard good covers further down the list.
        if (a.thumbnailAssetId != null) withCovers.add(a);
      }

      stack.addAll(childrenByParent[current] ?? const <String>[]);
    }

    withCovers.sort(_byRecencyThenId);
    summaries[folder.id] = FolderSummary(
      albumCount: albumCount,
      previewAlbums: withCovers.take(_previewLimit).toList(),
    );
  }

  return summaries;
}

bool isDescendant(List<SpaceAlbumFolder> folders, String candidateId, String ancestorId) {
  final index = _byId(folders);
  if (!index.containsKey(candidateId) || !index.containsKey(ancestorId) || candidateId == ancestorId) {
    return false;
  }
  return _ancestors(index, candidateId).skip(1).any((f) => f.id == ancestorId);
}

List<FolderSearchHit> flattenForSearch(List<SpaceAlbumFolder> folders, List<SpaceAlbum> albums, String query) {
  final needle = query.trim().toLowerCase();
  // A blank query means "search is inactive", not "match everything" — the caller uses a non-empty
  // query as the signal to switch into flattened mode at all.
  if (needle.isEmpty) return const [];

  final index = _byId(folders);
  // Name OR description, matching web's flattenForSearch and the flat filterAndSortSpaceAlbums
  // this path replaces while a query is active. SpaceAlbum.description is carried on the model for
  // exactly this, so leaving it out here silently narrows search compared with both.
  bool matches(SpaceAlbum a) =>
      a.name.toLowerCase().contains(needle) || (a.description ?? '').toLowerCase().contains(needle);

  return albums.where(matches).map((a) {
    final path = a.folderId == null ? <String>[] : _ancestors(index, a.folderId!).reversed.map((f) => f.name).toList();
    return FolderSearchHit(a, path);
  }).toList();
}

File: web/src/lib/utils/space-album-folders.ts
import type { SharedSpaceAlbumFolderDto, SharedSpaceLinkedAlbumDto } from '@immich/sdk';

export type FolderNode = {
  folder: SharedSpaceAlbumFolderDto;
  children: FolderNode[];
};

export type SearchHit = {
  album: SharedSpaceLinkedAlbumDto;
  path: string[];
};

const FOLDER_PREVIEW_LIMIT = 4;

/**
 * Every function here is total over the two arrays the albums page already holds. They must
 * never throw on inconsistent input: another editor can delete a folder between our folder
 * fetch and our album fetch, so a `parentId` pointing at a folder we never received is normal,
 * not exceptional. Such a folder is treated as a root.
 */
const byId = (folders: SharedSpaceAlbumFolderDto[]) => new Map(folders.map((f) => [f.id, f]));

/** Walks parent links upward, guarding against a cycle in malformed data. */
const ancestorsOf = (index: Map<string, SharedSpaceAlbumFolderDto>, folderId: string): SharedSpaceAlbumFolderDto[] => {
  const chain: SharedSpaceAlbumFolderDto[] = [];
  const seen = new Set<string>();
  let current = index.get(folderId);

  while (current && !seen.has(current.id)) {
    seen.add(current.id);
    chain.push(current);
    current = current.parentId ? index.get(current.parentId) : undefined;
  }

  return chain;
};

export const buildFolderTree = (folders: SharedSpaceAlbumFolderDto[]): FolderNode[] => {
  const index = byId(folders);
  const nodes = new Map<string, FolderNode>(folders.map((f) => [f.id, { folder: f, children: [] }]));
  const roots: FolderNode[] = [];

  for (const folder of folders) {
    const node = nodes.get(folder.id)!;
    // A self-referencing parentId, or one we never received, both resolve to "root".
    const parent =
      folder.parentId && folder.parentId !== folder.id && index.has(folder.parentId)
        ? nodes.get(folder.parentId)
        : undefined;

    if (parent) {
      parent.children.push(node);
    } else {
      roots.push(node);
    }
  }

  // A cycle among two or more folders (e.g. A.parentId = B, B.parentId = A — neither
  // self-referencing) gives every node in the cycle a "valid" parent above, so none of them is
  // ever pushed to `roots`, and the whole cycle silently vanishes from the returned tree. Sweep
  // reachability from the roots we do have, then promote anything unreached — severing it from
  // whatever cyclic parent it was attached to, so a promoted node doesn't also linger as a child
  // inside a structure that would recurse forever. Both passes are O(n): the sweep visits each
  // node once, and the total work across all `.filter()` calls is bounded by the total number of
  // child-list entries, which is at most n.
  const reached = new Set<string>();
  const stack = [...roots];
  while (stack.length > 0) {
    const node = stack.pop()!;
    if (reached.has(node.folder.id)) {
      continue;
    }
    reached.add(node.folder.id);
    stack.push(...node.children);
  }

  for (const folder of folders) {
    if (reached.has(folder.id)) {
      continue;
    }
    const node = nodes.get(folder.id)!;
    const cyclicParent = folder.parentId ? nodes.get(folder.parentId) : undefined;
    if (cyclicParent) {
      cyclicParent.children = cyclicParent.children.filter((c) => c.folder.id !== folder.id);
    }
    roots.push(node);
  }

  return roots;
};

export const getFolderPath = (
  folders: SharedSpaceAlbumFolderDto[],
  folderId: string | null,
): SharedSpaceAlbumFolderDto[] => {
  if (!folderId) {
    return [];
  }
  return ancestorsOf(byId(folders), folderId).reverse();
};

export const getFolderContents = (
  folders: SharedSpaceAlbumFolderDto[],
  albums: SharedSpaceLinkedAlbumDto[],
  folderId: string | null,
): { folders: SharedSpaceAlbumFolderDto[]; albums: SharedSpaceLinkedAlbumDto[] } => {
  const index = byId(folders);

  if (folderId !== null && !index.has(folderId)) {
    return { folders: [], albums: [] };
  }

  const isRoot = (folder: SharedSpaceAlbumFolderDto) =>
    !folder.parentId || folder.parentId === folder.id || !index.has(folder.parentId);

  // T-08 (mirrored from mobile): an album whose folderId names a folder we have not loaded yet
  // falls back to the ROOT rather than being hidden. Another editor can delete a folder — or, on
  // mobile, sync can deliver the album-link row before the folder row — between our folder fetch
  // and our album fetch; either way an album vanishing from the grid is the worse failure.
  const effectiveFolderId = (album: SharedSpaceLinkedAlbumDto): string | null =>
    album.folderId && index.has(album.folderId) ? album.folderId : null;

  return {
    // A self-referencing folder (parentId === id) is a root per `isRoot` above — it must not
    // also satisfy `f.parentId === folderId` and list itself as its own child (U-02's case).
    folders: folders.filter((f) => (folderId === null ? isRoot(f) : f.parentId === folderId && f.id !== folderId)),
    albums: albums.filter((a) => effectiveFolderId(a) === folderId),
  };
};

const subtreeIds = (folders: SharedSpaceAlbumFolderDto[], folderId: string): Set<string> => {
  const childrenByParent = new Map<string, string[]>();
  for (const folder of folders) {
    if (folder.parentId && folder.parentId !== folder.id) {
      childrenByParent.set(folder.parentId, [...(childrenByParent.get(folder.parentId) ?? []), folder.id]);
    }
  }

  const ids = new Set<string>();
  const queue = [folderId];
  while (queue.length > 0) {
    const current = queue.pop()!;
    if (ids.has(current)) {
      continue;
    }
    ids.add(current);
    queue.push(...(childrenByParent.get(current) ?? []));
  }
  return ids;
};

const albumsInSubtree = (
  folders: SharedSpaceAlbumFolderDto[],
  albums: SharedSpaceLinkedAlbumDto[],
  folderId: string,
): SharedSpaceLinkedAlbumDto[] => {
  const ids = subtreeIds(folders, folderId);
  return albums.filter((a) => a.folderId && ids.has(a.folderId));
};

export const getRecursiveAlbumCount = (
  folders: SharedSpaceAlbumFolderDto[],
  albums: SharedSpaceLinkedAlbumDto[],
  folderId: string,
): number => albumsInSubtree(folders, albums, folderId).length;

/** Recency key: an album's newest photo, falling back to its own updatedAt when undated. */
const recencyOf = (album: SharedSpaceLinkedAlbumDto): number =>
  new Date(album.endDate ?? album.updatedAt ?? 0).getTime();

/**
 * Newest first, ties broken by id.
 *
 * The tie-break is not cosmetic. Albums routinely share a recency — a bulk import lands many in
 * the same tick, and an album with no photos falls back to a timestamp it shares with its
 * siblings. Without a total order, which four covers a folder shows depends on the order the
 * albums happen to arrive in, so the same folder can render a different collage after a refetch,
 * and two callers that walk the space differently disagree about the same folder. Mobile's
 * `_byRecencyThenId` is the same comparator, for the same reason.
 */
const byRecencyThenId = (a: SharedSpaceLinkedAlbumDto, b: SharedSpaceLinkedAlbumDto): number =>
  recencyOf(b) - recencyOf(a) || a.id.localeCompare(b.id);

export const getFolderPreviewAssetIds = (
  folders: SharedSpaceAlbumFolderDto[],
  albums: SharedSpaceLinkedAlbumDto[],
  folderId: string,
): string[] =>
  albumsInSubtree(folders, albums, folderId)
    // A null cover means getLinkedAlbums found no space-visible asset. Emitting it would
    // render a broken tile, which is the exact bug the server-side COALESCE prevents.
    .filter((a) => a.albumThumbnailAssetId !== null)
    .sort(byRecencyThenId)
    .slice(0, FOLDER_PREVIEW_LIMIT)
    .map((a) => a.albumThumbnailAssetId as string);

export type FolderSummary = { albumCount: number; previewAssetIds: string[] };

export const EMPTY_FOLDER_SUMMARY: FolderSummary = { albumCount: 0, previewAssetIds: [] };

/**
 * Every folder's recursive album count and preview covers, in one pass over the space.
 *
 * `getRecursiveAlbumCount` and `getFolderPreviewAssetIds` are each O(folders + albums) on their
 * own — both rebuild the parent index and re-scan every album in the space. Calling them from a
 * card's props, once per rendered folder, makes a level of F folders cost O(F × (folders +
 * albums)) and re-pays it on every reactive change. At this feature's own limits (500 folders per
 * space, and `GET /:id/albums` deliberately unpaginated) that is millions of array operations per
 * render.
 *
 * Here the two indexes are built once and shared, so each folder only walks its own subtree:
 * O(folders × depth + albums) for the whole level, with depth capped at 10 server-side.
 *
 * Cycle and dangling-parent behaviour is deliberately identical to the single-folder functions —
 * self-references are not treated as children, and the `seen` set terminates any longer cycle,
 * with every member of it counting the others' albums exactly as `subtreeIds` does. The
 * "agrees with the single-folder functions" test pins that equivalence so the two cannot drift.
 */
export const buildFolderSummaries = (
  folders: SharedSpaceAlbumFolderDto[],
  albums: SharedSpaceLinkedAlbumDto[],
): Map<string, FolderSummary> => {
  const childrenByParent = new Map<string, string[]>();
  for (const folder of folders) {
    const parentId = folder.parentId;
    // A self-reference is not a child of itself — same rule as `subtreeIds`.
    if (!parentId || parentId === folder.id) {
      continue;
    }
    const siblings = childrenByParent.get(parentId);
    if (siblings) {
      siblings.push(folder.id);
    } else {
      childrenByParent.set(parentId, [folder.id]);
    }
  }

  const albumsByFolder = new Map<string, SharedSpaceLinkedAlbumDto[]>();
  for (const album of albums) {
    const folderId = album.folderId;
    if (!folderId) {
      continue;
    }
    const bucket = albumsByFolder.get(folderId);
    if (bucket) {
      bucket.push(album);
    } else {
      albumsByFolder.set(folderId, [album]);
    }
  }

  const summaries = new Map<string, FolderSummary>();
  const seen = new Set<string>();
  const stack: string[] = [];

  for (const folder of folders) {
    seen.clear();
    stack.length = 0;
    stack.push(folder.id);

    let albumCount = 0;
    const withCovers: SharedSpaceLinkedAlbumDto[] = [];

    while (stack.length > 0) {
      const current = stack.pop()!;
      if (seen.has(current)) {
        continue;
      }
      seen.add(current);

      for (const album of albumsByFolder.get(current) ?? []) {
        albumCount++;
        // A null cover means getLinkedAlbums found no space-visible asset; it still counts
        // toward the album total, but emitting it would render a broken tile.
        if (album.albumThumbnailAssetId !== null) {
          withCovers.push(album);
        }
      }

      const children = childrenByParent.get(current);
      if (children) {
        stack.push(...children);
      }
    }

    summaries.set(folder.id, {
      albumCount,
      previewAssetIds: withCovers
        .sort(byRecencyThenId)
        .slice(0, FOLDER_PREVIEW_LIMIT)
        .map((a) => a.albumThumbnailAssetId as string),
    });
  }

  return summaries;
};

export const isDescendant = (
  folders: SharedSpaceAlbumFolderDto[],
  candidateId: string,
  ancestorId: string,
): boolean => {
  const index = byId(folders);
  if (!index.has(candidateId) || !index.has(ancestorId) || candidateId === ancestorId) {
    return false;
  }
  return ancestorsOf(index, candidateId)
    .slice(1)
    .some((f) => f.id === ancestorId);
};

export const flattenForSearch = (
  folders: SharedSpaceAlbumFolderDto[],
  albums: SharedSpaceLinkedAlbumDto[],
  query: string,
): SearchHit[] => {
  const needle = (query ?? '').trim().toLowerCase();
  // A blank query means "search is inactive", not "match everything" — the caller uses a
  // non-empty query as the signal to switch into flattened mode at all.
  if (!needle) {
    return [];
  }

  const index = byId(folders);

  return albums
    .filter((a) => a.albumName.toLowerCase().includes(needle) || (a.description ?? '').toLowerCase().includes(needle))
    .map((album) => ({
      album,
      path: album.folderId
        ? ancestorsOf(index, album.folderId)
            .reverse()
            .map((f) => f.name)
        : [],
    }));
};
